filament/backend/src/vulkan/VulkanTexture.cpp

```diff
@@ -447,6 +474,17 @@ void VulkanTexture::transitionLayout(VkCommandBuffer commands, const VkImageSubr
     }
 }
 
+// Notifies the texture that a particular subresource's layout has changed.
+void VulkanTexture::trackLayout(uint32_t miplevel, uint32_t layer, VkImageLayout layout) {
+    const uint32_t first = (layer << 16) | miplevel;
+    const uint32_t last = (layer << 16) | (miplevel + 1);
```

Comment:
makes me a little nervous to see a `uint32_t` shifted by 16 into another `uint32_t`. Could you just use a `std::pair` or your own little structure?
If not, at least have an `assert()` that we don't have overflows, or use a `uint16_t` for `layer` and `miplevel`.

Reply:
Good catch, I'll add an assert_invariant here and elsewhere.  ( I sort of want to keep uint32_t in the func sig to be consistent with the Vulkan structures, which are fairly stable. )